src/decrypt.jl

```diff
@@ -1,38 +1,12 @@
-
 """
 Common interface for all modes and key lengths
 """
-function decrypt(ciphertext::AESCipherText, cipher::AESCipher; remove_pad=true)
-	if ciphertext.mode !== get_mode(cipher)
-		error("Mismatching mode and cipher")
-	end
-	if ciphertext.keylength !== get_key_length(cipher)
-		error("Mismatching keylength")
-	end
-	if iscbc(cipher)
-		raw = AESCBC_D(ciphertext.data, ciphertext.iv, cipher.key, cipher.cache; remove_pad=remove_pad)
-	elseif isctr(cipher)
-		raw = AESCTR_D(ciphertext.data, ciphertext.iv, cipher.key, cipher.cache)
-	elseif isecb(cipher)
-		raw = AESECB_D(ciphertext.data, cipher.key, cipher.cache)
-	end
-	return (ciphertext.original_type)(raw)
-end
+function decrypt end
 
-function decrypt!(plaintext, ciphertext::AESCipherText, cipher::AESCipher; remove_pad=true)
-	if ciphertext.mode !== get_mode(cipher)
-		error("Mismatching mode and cipher")
-	end
-	if ciphertext.keylength !== get_key_length(cipher)
-		error("Mismatching keylength")
-	end
-	if iscbc(cipher)
-		AESCBC_D!(plaintext, ciphertext.data, ciphertext.iv, cipher.key, cipher.cache; remove_pad=remove_pad)
-	elseif isctr(cipher)
-		AESCTR_D!(plaintext, ciphertext.data, ciphertext.iv, cipher.key, cipher.cache)
-	elseif isecb(cipher)
-		AESECB_D!(plaintext, ciphertext.data, cipher.key, cipher.cache)
-	end
-	return (ciphertext.original_type)(plaintext)
```

Comment:
I differ from the opinion that only ByteArray should be returned. But we can do one thing, we should probably make some kwarg like `to_original_type` which should be true by default. If true, make the type cast, otherwise return byte array

Reply:
Well, the problem with returning something other than a vector of bytes is that simply from the ciphertext it's generally impossible to know whether the resulting plaintext is valid UTF-8 or not. Imagine someone trying to decrypt some other text, doesn't know that all Strings in julia must be valid UTF-8 if created via `String` and says "well I want text". That will not work and the decryption itself failed and discards the result, instead of the data still being available to them.

Additionally, having that conversion as a field of the ciphertext leads to type instability, since there _has to be_ a dynamic lookup of the result, which would be good to avoid.

Reply:
Yeah that's a valid point, Type instability might creep in. I think still we should find some good solution to it, but as of now this one looks good to merge. We can make it a new issue.